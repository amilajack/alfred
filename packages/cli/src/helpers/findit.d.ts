declare module 'findit' {
  import { EventEmitter } from 'events';
  export default function findit(dir: string): EventEmitter;
}
