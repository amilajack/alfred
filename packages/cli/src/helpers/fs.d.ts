/// <reference types="node" />
import fs from 'fs';
export declare const copyFileAsync: typeof fs.copyFile.__promisify__;
export declare const writeFileAsync: typeof fs.writeFile.__promisify__;
export declare const statAsync: typeof fs.stat.__promisify__;
export declare const readFileAsync: typeof fs.readFile.__promisify__;
//# sourceMappingURL=fs.d.ts.map