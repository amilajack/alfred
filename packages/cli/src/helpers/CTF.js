import path from 'path';
import fs from 'fs';
import npm from 'npm';
import yarn from 'yarn-api';
import CTF, {
  getDevDependencies,
  CORE_CTFS,
  INTERFACE_STATES,
  normalizeInterfacesOfSkill,
  AddCtfHelpers
} from '@alfredpkg/core';
import type { CtfMap, InterfaceState } from '@alfredpkg/core';
import ValidateConfig from './Validation';
import { getProjectRoot } from './CLI';

export const ENTRYPOINTS = [
  'lib.node.js',
  'app.node.js',
  'lib.browser.js',
  'app.browser.js',
  'lib.electron.main.js',
  'lib.electron.renderer.js',
  'app.electron.main.js',
  'app.electron.renderer.js',
  'lib.react-native.js',
  'app.react-native.js'
];

const projectRoot = getProjectRoot();

export function generateInterfaceStatesFromProject(): Array<InterfaceState> {
  const envs = ['production', 'development', 'test'];
  // Default to development env if no config given
  const env = envs.includes(process.env.NODE_ENV)
    ? process.env.NODE_ENV
    : 'development';

  return ENTRYPOINTS.filter(e =>
    fs.existsSync(path.join(projectRoot, 'src', e))
  ).map(e => {
    const [projectType, target] = e.split('.');
    return {
      env,
      target,
      projectType
    };
  });
}

/**
 * Find all the dependencies that are different between two CTF's.
 * This is used to figure out which deps need to be installed
 */
export function diffCtfDeps(oldCtf: CtfMap, newCtf: CtfMap): Array<string> {
  // Find the dependencies that have changed and install them
  const t: Map<string, string> = new Map();
  const s: Map<string, string> = new Map();

  Object.entries(getDevDependencies(oldCtf)).forEach(([key, val]) => {
    t.set(key, val);
  });
  Object.entries(getDevDependencies(newCtf)).forEach(([key, val]) => {
    if (t.has(key)) {
      if (t.get(key) !== val) {
        throw new Error('Cannot resolve diff deps');
      }
    } else {
      s.set(key, val);
    }
  });

  return Array.from(s.entries()).map(([key, val]) => `${key}@${val}`);
}

/**
 * @TODO Account for `devDependencies` and `dependencies`
 */
export function installDeps(
  dependencies: Array<string> = [],
  npmClient: 'npm' | 'yarn' = 'npm'
): Promise<any> {
  if (!dependencies.length) return Promise.resolve();

  switch (npmClient) {
    case 'npm': {
      return new Promise((resolve, reject) => {
        npm.load({ save: true }, err => {
          if (err) reject(err);

          npm.commands.install(dependencies, (_err, data) => {
            if (_err) reject(_err);
            resolve(data);
          });

          npm.on('log', console.log);
        });
      });
    }
    case 'yarn': {
      return new Promise((resolve, reject) => {
        yarn(['why', 'isobject'], err => {
          if (err) reject(err);
          resolve();
        });
      });
    }
    default: {
      throw new Error('Unsupported npm client. Can only be "npm" or "yarn"');
    }
  }
}

export type AlfredConfig = {
  npmClient: 'npm' | 'yarn',
  skills: Array<string>,
  root: string,
  showConfigs: boolean
};

/**
 * Add skills to a given list of skills to ensure that the list has a complete set
 * of standard ctfs
 */
export function addMissingStdSkillsToCtf(
  ctf: CtfMap,
  alfredConfig: AlfredConfig,
  interfaceState: InterfaceState
): CtfMap {
  const stdCtf = new Map(
    Object.entries({
      lint: CORE_CTFS.eslint,
      format: CORE_CTFS.prettier,
      build: require('@alfredpkg/interface-build').resolveSkill(
        Object.values(CORE_CTFS),
        interfaceState
      ),
      start: require('@alfredpkg/interface-start').resolveSkill(
        Object.values(CORE_CTFS),
        interfaceState
      ),
      test: CORE_CTFS.jest
    })
  );

  ctf.forEach(ctfNode => {
    /* eslint no-param-reassign: off */
    ctfNode.interfaces = normalizeInterfacesOfSkill(ctfNode.interfaces);
  });

  const stdSubCommands: Set<string> = new Set(stdCtf.keys());
  // Create a set of subcommands that the given CTF has
  const ctfSubcommands: Set<string> = Array.from(ctf.values()).reduce(
    (prev, ctfNode) => {
      if (ctfNode.interfaces && ctfNode.interfaces.length) {
        ctfNode.interfaces.forEach(_interface => {
          const { subcommand } = _interface.module;
          prev.add(subcommand);
        });
      }
      return prev;
    },
    new Set()
  );

  stdSubCommands.forEach(stdSubCommand => {
    if (!ctfSubcommands.has(stdSubCommand)) {
      const stdCtfSkillToAdd = stdCtf.get(stdSubCommand);
      if (
        stdCtfSkillToAdd &&
        stdCtfSkillToAdd.interfaces &&
        stdCtfSkillToAdd.interfaces.length
      ) {
        stdCtfSkillToAdd.interfaces = normalizeInterfacesOfSkill(
          stdCtfSkillToAdd.interfaces
        );
      }
      ctf.set(stdCtfSkillToAdd.name, {
        ...stdCtfSkillToAdd,
        ...AddCtfHelpers
      });
    }
  });

  // Add all the CORE_CTF's without subcommands
  ctf.set('babel', { ...CORE_CTFS.babel, ...AddCtfHelpers });
  // @TODO
  // ctf.set('lodash', { ...CORE_CTFS.lodash, ...AddCtfHelpers });

  ctf.forEach(ctfNode => {
    const ctfWithHelpers = {
      ...ctfNode,
      ...AddCtfHelpers
    };
    Object.entries(ctfWithHelpers.ctfs || {}).forEach(([ctfName, ctfFn]) => {
      const correspondingCtfNode = ctf.get(ctfName);
      if (correspondingCtfNode) {
        ctf.set(
          ctfName,
          ctfFn(correspondingCtfNode, ctf, { alfredConfig, ...interfaceState })
        );
      }
    });
  });

  return ctf;
}

export async function loadConfigs(
  pkgPath: string = path.join(projectRoot, 'package.json')
): Promise<{ pkg: Object, pkgPath: string, alfredConfig: AlfredConfig }> {
  if (!fs.existsSync(pkgPath)) {
    throw new Error('Current working directory does not have "package.json"');
  }

  // Read the package.json and validate the Alfred config
  const pkg = JSON.parse((await fs.promises.readFile(pkgPath)).toString());
  const tmpAlfredConfig = pkg.alfred || {};
  ValidateConfig(tmpAlfredConfig || {});

  const defaultOpts = {
    npmClient: 'npm',
    skills: [],
    root: projectRoot
  };
  const alfredConfig = Object.assign({}, defaultOpts, tmpAlfredConfig);

  return { pkg, pkgPath, alfredConfig };
}

export default async function generateCtfFromConfig(
  alfredConfig: AlfredConfig,
  interfaceState: InterfaceState
): Promise<CtfMap> {
  // Check if any valid entrypoints exist
  const states = generateInterfaceStatesFromProject();
  if (!states.length) {
    throw new Error(
      `The project must have at least one entrypoint. Here are some examples of entrypoints:\n\n${ENTRYPOINTS.map(
        e => `"./src/${e}"`
      ).join('\n')}`
    );
  }

  // Generate the CTF
  const tmpCtf: CtfMap = new Map();
  const { skills = [] } = alfredConfig;
  skills.forEach(skill => {
    /* eslint-disable */
    const c = require(skill);
    /* eslint-enable */
    tmpCtf.set(c.name, c);
  });

  const ctf = CTF(Array.from(tmpCtf.values()), alfredConfig, interfaceState);
  addMissingStdSkillsToCtf(ctf, alfredConfig, interfaceState);

  return ctf;
}

export async function diffCtfDepsOfAllInterfaceStates(
  prevAlfredConfig: AlfredConfig,
  currAlfredConfig: AlfredConfig
): Set<string> {
  const stateWithDuplicateDeps = await Promise.all(
    INTERFACE_STATES.map(state =>
      Promise.all([
        generateCtfFromConfig(prevAlfredConfig, state),
        generateCtfFromConfig(currAlfredConfig, state)
      ])
    )
  );

  return Array.from(
    new Set(
      stateWithDuplicateDeps
        .map(([a, b]) => diffCtfDeps(a, b))
        .reduce((prev, curr) => prev.concat(curr), [])
    )
  );
}
