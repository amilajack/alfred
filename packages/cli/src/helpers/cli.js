// @flow
import path from 'path';
import fs from 'fs';
import pkgUp from 'pkg-up';
import rimraf from 'rimraf';
import { getConfigsBasePath, loadConfig, formatPkgJson } from '@alfred/core';
import type { AlfredConfig, InterfaceState } from '@alfred/core';
import handlebars from 'handlebars';
import { Signale } from 'signale';
import { ENTRYPOINTS, generateInterfaceStatesFromProject } from './ctf';
import PkgValidation from '../pkg-validation';

export const getInstallCommmand = (alfredConfig: AlfredConfig): string => {
  const { root, npmClient } = alfredConfig;
  return npmClient.toLowerCase() === 'npm'
    ? `npm install --prefix ${root}`
    : 'yarn';
};

/**
 * Execute promises serially
 */
export const serial = (funcs: Array<() => Promise<any>>) =>
  funcs.reduce(
    (promise, func) =>
      // eslint-disable-next-line promise/no-nesting
      promise.then(result => func().then(Array.prototype.concat.bind(result))),
    Promise.resolve([])
  );

export function validatePkg() {
  const pkgPath = path.join(process.cwd(), 'package.json');
  const result = PkgValidation.validate(fs.readFileSync(pkgPath).toString());

  if (result.messagesCount) {
    const signale = new Signale();
    signale.note(pkgPath);
    result.recommendations.forEach(warning => {
      signale.warn(warning);
    });
    result.warnings.forEach(warning => {
      signale.warn(warning);
    });
    result.errors.forEach(warning => {
      signale.error(warning);
    });
  }
}

/**
 * Check if a directory contains an Alfred project
 */
export function checkIsAlfredProject(
  config: AlfredConfig,
  interfaceStates: Array<InterfaceState>
) {
  const srcPath = path.join(config.root, 'src');
  validatePkg();

  if (!fs.existsSync(srcPath)) {
    throw new Error(
      '"./src" directory does not exist. This does not seem to be an Alfred project'
    );
  }

  const hasEntrypoint = ENTRYPOINTS.some(e =>
    fs.existsSync(path.join(srcPath, e))
  );

  if (!hasEntrypoint) {
    throw new Error(
      `You might be in the wrong directory or this is not an Alfred project. The project must have at least one entrypoint. Here are some examples of entrypoints:\n\n${ENTRYPOINTS.map(
        e => `"./src/${e}"`
      ).join('\n')}`
    );
  }

  // Run validation that is specific to each interface state
  interfaceStates
    .map(interfaceState =>
      [
        interfaceState.projectType,
        interfaceState.target,
        interfaceState.env
      ].join('.')
    )
    .forEach(interfaceStateString => {
      switch (interfaceStateString) {
        case 'app.browser.production':
        case 'app.browser.development': {
          const indexHtmlPath = path.join(srcPath, 'index.html');
          if (!fs.existsSync(indexHtmlPath)) {
            throw new Error(
              'An "./src/index.html" file must exist when targeting a browser environment'
            );
          }
          break;
        }
        default:
          break;
      }
    });
}

/**
 * Get the root of a project from the current working directory
 */
export function getProjectRoot() {
  const pkgPath = pkgUp.sync();
  if (!pkgPath) {
    throw new Error(
      `Alfred project root could not be found from "${process.cwd()}".

      Make sure you are inside an Alfred project.`
    );
  }
  return path.dirname(pkgPath);
}

const TEMPLATES_DIR = path.resolve(__dirname, '../templates');

async function compileTemplate(templateFilename: string) {
  const source = await fs.promises.readFile(
    path.resolve(TEMPLATES_DIR, templateFilename)
  );
  return handlebars.compile(source.toString(), { noEscape: true });
}

/**
 * @param {*} templateData
 * @param {*} root
 * @param {*} entrypointInterfaceStates These come in the form of [['lib', 'node], ...etc]
 */
export async function addEntrypoints(
  rawTemplateData: Object,
  root: string,
  entrypointInterfaceStates: Array<InterfaceState>
) {
  const [
    APP_TEMPLATE,
    LIB_TEMPLATE,
    APP_BROWSER_HTML_TEMPLATE,
    TEST_TEMPLATE
  ] = await Promise.all(
    ['app.js.hbs', 'lib.js.hbs', 'index.html.hbs', 'test.js.hbs'].map(
      compileTemplate
    )
  );

  await Promise.all(
    entrypointInterfaceStates.map(interfaceState => {
      const { projectType, target } = interfaceState;
      const isApp = projectType === 'app';
      const isBrowser = target === 'browser';

      const templateData = {
        ...rawTemplateData,
        project: { projectType, target, isApp, isBrowser }
      };

      const writeIndexHtml = () => {
        if (isApp && isBrowser) {
          const content = APP_BROWSER_HTML_TEMPLATE(templateData);
          return fs.promises.writeFile(
            path.join(root, './src/index.html'),
            content
          );
        }
        return Promise.resolve();
      };

      return Promise.all(
        [
          {
            file: `./src/${projectType}.${target}.js`,
            content: (isApp ? APP_TEMPLATE : LIB_TEMPLATE)(templateData)
          },
          {
            file: `./tests/${projectType}.${target}.spec.js`,
            content: TEST_TEMPLATE(templateData)
          }
        ]
          .map(({ file, content }) =>
            fs.promises.writeFile(path.join(root, file), content)
          )
          .concat([writeIndexHtml()])
      );
    })
  );
}

// This function lives here because it is used in both tests and the cli
export async function addBoilerplate(templateData: Object, root: string) {
  const [
    GITIGNORE_TEMPLATE,
    NPM_TEMPLATE,
    README_TEMPLATE,
    EDITORCONFIG_TEMPLATE
  ] = await Promise.all(
    [
      '.gitignore.hbs',
      'package.json.hbs',
      'README.md.hbs',
      '.editorconfig.hbs'
    ].map(compileTemplate)
  );

  await Promise.all(
    [
      {
        file: '.gitignore',
        content: GITIGNORE_TEMPLATE(templateData)
      },
      {
        file: '.editorconfig',
        content: EDITORCONFIG_TEMPLATE(templateData)
      },
      {
        file: 'README.md',
        content: README_TEMPLATE(templateData)
      }
    ].map(({ file, content }) =>
      fs.promises.writeFile(path.join(root, file), content)
    )
  );

  const pkgPath = path.join(root, 'package.json');
  const formattedPkg = await formatPkgJson(
    JSON.parse(NPM_TEMPLATE(templateData))
  );
  await fs.promises.writeFile(pkgPath, formattedPkg);
  const { projectType, target } = templateData.project;

  await addEntrypoints(templateData, root, [
    { projectType, target, env: 'production' }
  ]);
}

export async function init() {
  const projectRoot = getProjectRoot();
  const config = await loadConfig(projectRoot);
  const { alfredConfig } = config;
  const interfaceStates = generateInterfaceStatesFromProject(alfredConfig);
  checkIsAlfredProject(alfredConfig, interfaceStates);
  return { ...config, projectRoot };
}

/**
 * Delete .configs dir
 */
export function deleteConfigs(config: AlfredConfig): Promise<void> {
  const configsBasePath = getConfigsBasePath(config.root);
  if (fs.existsSync(configsBasePath)) {
    return new Promise(resolve => {
      rimraf(configsBasePath, () => {
        resolve();
      });
    });
  }
  return Promise.resolve();
}

export default function getSingleSubcommandFromArgs(
  args: Array<string>
): string {
  switch (args.length) {
    case 0: {
      throw new Error('One subcommand must be passed');
    }
    case 1: {
      break;
    }
    default: {
      throw new Error('Only one subcommand can be passed');
    }
  }

  return args[0];
}
