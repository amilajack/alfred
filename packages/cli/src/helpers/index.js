export * from './ctf';
export * from './cli';
export * from './parse-input';
