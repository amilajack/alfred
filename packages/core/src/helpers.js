/**
 * Execute promises serially
 * @REFACTOR There's cleaner ways of implementing this
 */
export const serial = (funcs: Array<() => Promise<any>>) =>
  funcs.reduce(
    (promise, func) =>
      // eslint-disable-next-line promise/no-nesting
      promise.then(result => func().then(Array.prototype.concat.bind(result))),
    Promise.resolve([])
  );
