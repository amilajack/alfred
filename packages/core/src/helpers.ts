export function requireModule(moduleName: string): any {
  const requiredModule = require(moduleName);
  return requiredModule.default || requiredModule;
}
