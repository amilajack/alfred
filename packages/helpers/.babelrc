{
  "presets": [
    [
      "@babel/preset-env",
      {
        "targets": {
          "node": 13
        }
      }
    ]
  ],
  "plugins": [
    "@babel/plugin-syntax-dynamic-import",
    "@babel/plugin-proposal-class-properties",
    "@babel/plugin-proposal-export-namespace-from"
  ]
}
