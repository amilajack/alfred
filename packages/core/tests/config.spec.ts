import { requireConfig } from '@alfred/helpers';
import Config from '../src/config';

describe('config', () => {
  describe('config', () => {
    it('should take plain object', () => {
      expect(
        new Config({
          autoInstall: false
        }).getConfigValues()
      ).toEqual({
        ...Config.DEFAULT_CONFIG,
        autoInstall: false
      });
    });

    it('should take an object with empty extends', () => {
      expect(
        new Config({
          extends: []
        }).getConfigValues()
      ).toEqual({ ...Config.DEFAULT_CONFIG });
    });

    it('should accept strings in ".extends" array', () => {
      jest.mock(
        'module-1',
        () => ({
          showConfigs: true
        }),
        { virtual: true }
      );
      jest.mock(
        'module-2',
        () => ({
          showConfigs: false
        }),
        { virtual: true }
      );
      expect(
        new Config({
          extends: 'module-1'
        }).getConfigValues()
      ).toEqual({
        ...Config.DEFAULT_CONFIG,
        showConfigs: true
      });
      expect(
        new Config({
          extends: ['module-2'],
          showConfigs: true
        }).getConfigValues()
      ).toEqual({
        ...Config.DEFAULT_CONFIG,
        showConfigs: true
      });
    });

    it('should throw if extends property is not a string or an array', () => {
      expect(
        () =>
          new Config({
            // @ts-ignore
            // eslint-disable-next-line @typescript-eslint/no-empty-function
            extends: (): void => {}
          })
      ).toThrowErrorMatchingSnapshot();
    });

    it('should not accept objects as input', () => {
      expect(
        () =>
          new Config({
            // @ts-ignore
            extends: [{ extends: {} }]
          })
      ).toThrow();
    });

    it('should require non-prefixed modules', () => {
      jest.mock(
        'alfred-config-bliss',
        () => ({
          autoInstall: true,
          skills: ['@alfred/skill-react']
        }),
        { virtual: true }
      );
      expect(
        new Config({
          extends: ['bliss']
        }).getConfigValues()
      ).toEqual({
        ...Config.DEFAULT_CONFIG,
        autoInstall: true,
        skills: [['@alfred/skill-react', {}]]
      });
      expect(
        new Config({
          extends: 'alfred-config-bliss'
        }).getConfigValues()
      ).toEqual({
        ...Config.DEFAULT_CONFIG,
        autoInstall: true,
        skills: [['@alfred/skill-react', {}]]
      });
    });

    it('should accept options that can be overriden', () => {
      jest.mock(
        'alfred-config-test',
        () => ({
          showConfigs: true,
          skills: [
            [
              '@alfred/skill-lodash',
              {
                collections: true,
                paths: true
              }
            ]
          ]
        }),
        { virtual: true }
      );
      expect(
        new Config({
          extends: 'alfred-config-test'
        }).getConfigValues()
      ).toEqual({
        ...Config.DEFAULT_CONFIG,
        showConfigs: true,
        skills: [
          [
            '@alfred/skill-lodash',
            {
              collections: true,
              paths: true
            }
          ]
        ]
      });
      expect(
        new Config({
          extends: 'alfred-config-test',
          skills: [
            '@alfred/skill-parcel',
            [
              '@alfred/skill-babel',
              {
                plugins: ['@babel/preset-flow']
              }
            ],
            [
              '@alfred/skill-babel',
              {
                plugins: ['@babel/preset-react']
              }
            ]
          ]
        }).getConfigValues()
      ).toEqual({
        ...Config.DEFAULT_CONFIG,
        showConfigs: true,
        skills: [
          [
            '@alfred/skill-lodash',
            {
              collections: true,
              paths: true
            }
          ],
          ['@alfred/skill-parcel', {}],
          [
            '@alfred/skill-babel',
            {
              plugins: ['@babel/preset-flow', '@babel/preset-react']
            }
          ]
        ]
      });
    });

    it('should throw if require non-existent module', () => {
      expect(() => requireConfig('foo')).toThrowErrorMatchingSnapshot();
    });
  });
});
