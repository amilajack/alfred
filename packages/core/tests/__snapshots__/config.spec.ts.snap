// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config config should throw if extends property is not a string or an array 1`] = `"Values in \\".extends\\" property in Alfred config must be a string. Instead passed () => {}"`;

exports[`config config should throw if require non-existent module 1`] = `"Could not resolve \\"foo\\" module or \\"eslint-config-foo\\" module"`;
